src/python/tests/gff3/test_extract_annotation.py

```diff
@@ -22,15 +22,230 @@
+            assert annot.features[feat_type][feature.id]
+
+    @pytest.mark.parametrize(
+        "parent_type, parent_id, child_id, expected",
+        [
+            ("gene", "geneA", "mrnA", does_not_raise()),
+            ("bad_type", "geneA", "mrnA", raises(KeyError)),
+            ("gene", "geneB", "mrnA", raises(MissingParentError)),
+        ],
+    )
+    def test_add_parent(self, parent_type, parent_id, child_id, expected):
+        """Tests the `FunctionaAnnotation.add_parent_link()` method.
+
+        Add a parent feature, and then add a parent link.
+
+        Args:
+            parent_type: Type for the SeqFeature parent.
+            parent_id: ID for the SeqFeature parent.
+            child_id: ID for the SeqFeature child.
+            expected: What exception is expected to be raised, if any.
+
+        """
+        annot = FunctionalAnnotations()
+        parent = SeqFeature(type="gene", id="geneA")
+        annot.add_feature(parent, "gene")
+
+        with expected:
+            annot.add_parent_link(parent_type, parent_id, child_id)
+
+    @pytest.mark.parametrize(
+        "in_parent_type, in_parent_id, in_child_id, out_parent_type, out_child_id, expected",
+        [
+            ("gene", "geneA", "mrnA", "gene", "mrnA", does_not_raise()),
+            ("gene", "geneA", "mrnA", "bad_type", "mrnA", raises(KeyError)),
+            ("gene", "geneA", "mrnA", "gene", "mrnB", raises(MissingParentError)),
+        ],
+    )
+    def test_get_parent(
+        self, in_parent_type, in_parent_id, in_child_id, out_parent_type, out_child_id, expected
+    ):
+        """Tests the `FunctionaAnnotation.get_parent()` method.
+
+        Args:
+            in_parent_type: Type for the SeqFeature parent.
+            in_parent_id: ID for the SeqFeature parent.
+            in_child_id: ID for the SeqFeature child.
+            out_parent_type: Type for the parent stored in the FunctionalAnnotation.
+            out_parent_id: ID for the parent stored.
+            expected: What exception is expected to be raised, if any.
+
+        """
+        annot = FunctionalAnnotations()
+        parent = SeqFeature(type=in_parent_type, id=in_parent_id)
+        annot.add_feature(parent, "gene")
+        annot.add_feature(
+            SeqFeature(type="mRNA", id=in_child_id), feat_type="transcript", parent_id=in_parent_id
+        )
+
+        with expected:
+            out_parent = annot.get_parent(out_parent_type, out_child_id)
+            assert out_parent == in_parent_id
+
+    @pytest.mark.parametrize(
+        "child_type, child_id, out_parent_id, expected",
+        [
+            ("transcript", "mrna_A", "gene_A", does_not_raise()),
+            ("bad_type", "mrna_A", "gene_A", raises(KeyError)),  # Child type does not exist
+            ("gene", "gene_A", None, raises(AnnotationError)),  # Feature ID already loaded
+            ("gene", "gene_B", "gene_A", raises(AnnotationError)),  # Cannot add a gene child of a gene
```

Comment:
Alternative proposal:
```suggestion
            ("bad_type", "mrna_A", "gene_A", raises(KeyError), id="unknown_child_type"),
            ("gene", "gene_A", None, raises(AnnotationError), id="feature_id_already_loaded"),
            ("gene", "gene_B", "gene_A", raises(AnnotationError), id="circular_gene_addition"),
```